tensorflow/core/kernels/stack.cc

```diff
@@ -231,52 +319,64 @@ void StackPushOp::ComputeAsync(OpKernelContext* ctx, DoneCallback done) {
-    Allocator* allocator = device->GetAllocator(alloc_attrs);
-    absl::optional<AllocatorStats> stats = allocator->GetStats();
-    if (stats && *stats->bytes_limit &&
-        stats->bytes_in_use > (*stats->bytes_limit * kOccupancy)) {
-      // Asynchronously copy the tensor from GPU to CPU memory.
-      // TODO(yuanbyu): Swap the oldest tensor first.
-      AllocatorAttributes host_alloc_attrs;
-      host_alloc_attrs.set_gpu_compatible(true);
-      host_alloc_attrs.set_on_host(true);
-      Allocator* cpu_allocator = device->GetAllocator(host_alloc_attrs);
-      Tensor* cpu_tensor =
-          new Tensor(cpu_allocator, tensor.dtype(), tensor.shape());
-      device_ctxt->CopyDeviceTensorToCPU(
-          &tensor, "StackPush", device, cpu_tensor,
-          [cpu_tensor, stack, ctx, done](const Status& s) {
-            ctx->SetStatus(s);
-            if (s.ok()) {
-              AllocatorAttributes alloc_attrs = ctx->input_alloc_attr(1);
-              ctx->SetStatus(stack->Push({*cpu_tensor, alloc_attrs, true}));
-            }
-            if (ctx->status().ok()) {
-              ctx->set_output(0, *cpu_tensor);
-            }
-            done();
-            delete cpu_tensor;
-          });
-      return;
-    }
+
+  if (!swap_memory_ || alloc_attrs.on_host()) {
+    done();
+    return;
   }
 
-  // Execute synchronously if not swapped.
-  OP_REQUIRES_OK_ASYNC(ctx, stack->Push({tensor, alloc_attrs, false}), done);
-  ctx->set_output(0, tensor);
-  done();
+  DeviceContext* device_ctxt = ctx->op_device_context();
+  auto device = static_cast<tensorflow::Device*>(ctx->device());
+  Allocator* allocator = device->GetAllocator(alloc_attrs);
+  absl::optional<AllocatorStats> stats = allocator->GetStats();
+  if (!stats || !*stats->bytes_limit ||
+      stats->bytes_in_use <= (*stats->bytes_limit * kOccupancy)) {
+    done();
+    return;
+  }
+
+  // Obtain the oldest unswapped TensorAndAllocation.
+  Stack::TensorAndAllocation* to_swap_out = nullptr;
+  if (!stack->GetTensorToSwapOut(
+          [stack](const Stack::TensorAndAllocation& value) -> bool {
+            return value.tensor.TotalBytes() > kCopyThreshold &&
+                   stack->IsUsefulToSwap(value.tensor);
+          },
+          &to_swap_out)) {
+    done();
+    return;
+  }
+
+  // Asynchronously copy the tensor from GPU to CPU memory.
+  // Swap the oldest tensor first.
+  AllocatorAttributes host_alloc_attrs;
+  host_alloc_attrs.set_gpu_compatible(true);
+  host_alloc_attrs.set_on_host(true);
+  Allocator* cpu_allocator = device->GetAllocator(host_alloc_attrs);
+  Tensor* cpu_tensor = new Tensor(cpu_allocator, (to_swap_out->tensor).dtype(),
+                                  (to_swap_out->tensor).shape());
+  device_ctxt->CopyDeviceTensorToCPU(
+      &(to_swap_out->tensor), "StackPush", device, cpu_tensor,
+      [stack, to_swap_out, cpu_tensor, done](const Status& s) {
+        if (s.ok()) {
+          stack->MarkAsSwapped(to_swap_out, cpu_tensor);
```

Comment:
` error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]`

Reply:
I'll change the method return value to void, then.

Reply:
Done so with the 2nd latest commit.